subprojects/provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/DefaultProjectSchemaProvider.kt

```diff
@@ -92,7 +92,7 @@ fun accessibleConventionsSchema(plugins: Map<String, Any>) =
 
 private
 fun sourceSetsOf(project: Project) =
-    project.convention.findPlugin(JavaPluginConvention::class.java)?.sourceSets
+    project.extensions.findByType(typeOf<NamedDomainObjectContainer<SourceSet>>())
```

Comment:
:x: `project.extensions["sourceSets"] as SourceSetContainer`?

Reply:
That would break previous semantics and fail on projects that don't have the `sourceSets` extension.

Reply:
:x: I don't think finding by type is the correct thing to do here. We know we want to access the extension named `sourceSets` and we know its type.

Reply:
Right. What about:
```kotlin
project.extensions.findByName("sourceSets") as? SourceSetContainer
```
Using the indexer would throw.